### ISSUE TITLE
[Bug] Integration fails with Electric Eel - chart error

### ISSUE THREAD

**Issue by Berserkir-Wolf** [tomaae/homeassistant-truenas#170]
## Describe the issue
Using the beta version of TrueNAS-SCALE, the API for 'chart/release' fails with: 
`TrueNAS 192.168.10.11 unable to fetch data "chart/release" (404)`
This persists through reboots of both systems - it appears to be related to an API change in Electric Eel, likely relating to the changes to how containers are handled.

Generating a new API key also has the same error, which suggests it's not credential validation causing it, but likely an endpoint has changed. Looking further into the repo and the API itself, it is to do with the app releases section specifically. Bypassing that gets the normal data loading properly, just minus the container/app information.
(changing line 1123 in coordinator.py to remove the 'not' part allows for regular data to resume flow).

## How to reproduce the issue
1. Upgrade to 'Electric Eel' (beta)
2. Restart HAss
3. See error (unavailable entities)

Alternately:
Attempt to re-add to HAss, and get the specific error.

## Expected behavior
TrueNAS information shows up in HA.

## Software versions

 - Home Assistant version: 
 - TrueNAS integration version: 
 - TrueNAS version:

**Comment by awptechnologies** [comment]
Same issue here after the electric eel update this integration no longer works. I reinstalled it and it accepted my host and api key but it fails to load. Worked fine until new truenas update train.

**Comment by awptechnologies** [comment]
how do i make this fix and add it back to home assistant with the fix?

**Comment by Berserkir-Wolf** [comment]
> how do i make this fix and add it back to home assistant with the fix?

You need to alter the 'coordinator.py' file in the custom components folder. If you installed using HACS, the file is located in `custom_components/truenas/coordinator.py`. 

Line 1123 normally reads:
`        if not self._is_scale:`

If you remove the "not" part, it should read:
`        if self._is_scale:`

Then restart HA to load the modified version. It's not a fix, but it gets most stats back online until a proper fix is available (which I expect will require more changes from TrueNAS to finish the apps API on the new version).

**Comment by Berserkir-Wolf** [comment]
Commenting to keep this open for monitoring. RC of truenas releases in the next few days, so will see if the issue persists in the Release Candidate version.

**Comment by debsahu** [comment]
> Commenting to keep this open for monitoring. RC of truenas releases in the next few days, so will see if the issue persists in the Release Candidate version.

Issue is still there in RC1

**Comment by Berserkir-Wolf** [comment]
Confirmed, bug remains - removing the 'bypass' for the check causes the issue to reoccur.

**Comment by lyricnz** [comment]
Confirm. Updated truenas to EE RC1, which broke the integration.  Lots of errors like

```
2024-10-03 09:39:54.577 ERROR (MainThread) [homeassistant.util.logging] Exception in async_update_controller when dispatching
 'update_sensors': (<custom_components.portainer.coordinator.PortainerCoordinator object at 0x7fbf5c04dfd0>,)
Traceback (most recent call last):
  File "/config/custom_components/truenas/entity.py", line 68, in async_update_controller
    data = coordinator.data[entity_description.data_path]
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'system_info'
```

and KeyError: 'pool'

**Comment by tomaae** [comment]
No surprise there, since they are moving from k3s to docker. API documentation is not available right now, so probably after release we can get it updated to work with EE.
You will probably make it work by hashing out apps related updates in main coordinator thread.

**Comment by Berserkir-Wolf** [comment]
Yeah, or changing the scale check as I mentioned earlier so it just doesn't do those particular checks.
Given it's now a Release Candidate version, I'd say API docs shouldn't be too far away - RC should be feature complete, in theory.

**Comment by mstepuch** [comment]
Error still there

**Comment by mcouture87** [comment]
issue still exists in RC2

**Comment by ISilvaPT** [comment]
Electric Eel now final, if you can @tomaae please make it compatible 👍 .

**Comment by itsniawan** [comment]
if you already update to 24.10, you can make it works by editing few lines like i did.
in `custom_components/truenas/coordinator.py`, if you already edit line 1123 before, revert it back to 
```
        if not self._is_scale:
```
and edit line 1128 and so on to:

```
            source=self.api.query("app"),
            key="id",
            vals=[
                {"name": "id", "default": 0},
                {"name": "name", "default": "unknown"},
                {"name": "human_version", "default": "unknown"},
                {"name": "upgrade_available", "default": "unknown"},
                {"name": "image_updates_available", "default": "unknown"},
                {"name": "portal", "source": "portals/Web UI", "default": "unknown"},
                {"name": "state", "default": "unknown"},
            ],
            ensure_vals=[
                {"name": "running", "type": "bool", "default": False},
            ],
        )

        for uid, vals in self.ds["app"].items():
            self.ds["app"][uid]["running"] = vals["state"] == "RUNNING"
```

and for fixing the attibutes:
in `custom_components/truenas/binary_sensor_types.py` line 78-79
```
    "upgrade_available",
    "image_updates_available",
```

don't forget to restart homeassistant

![image](https://github.com/user-attachments/assets/73364fac-a42b-4cf2-b100-8077cf696ae6)

**Comment by Berserkir-Wolf** [comment]
That works - mostly. 
Still randomly suggests that there's a release update available, which is bizarre.
To make that change reliable while retaining support for people who haven't yet updated, we might need a way to check the version and see if it's pre- or post- changes, so that the API calls differ based on truenas version being connected to.

**Comment by Poshy163** [comment]
for sure, i can hack something together within the next few days if no one else has already started. 

got a feeling it might be a little more complex than first anticipated tho 🥲

**Comment by kmanan** [comment]
No response from host. Updated to Electric Eel

**Comment by itsniawan** [comment]
> That works - mostly. Still randomly suggests that there's a release update available, which is bizarre. To make that change reliable while retaining support for people who haven't yet updated, we might need a way to check the version and see if it's pre- or post- changes, so that the API calls differ based on truenas version being connected to.

I think i have found solution for this problems. I was trying to make a pull request but it was a mess so i delete it  (i'm still newbie 😅).
You can check all changes i made to make it works in my forked repo [itsniawan/homeassistant-truenas](https://github.com/itsniawan/homeassistant-truenas)

**Comment by Poshy163** [comment]
i think there is 2 ways this can go about, either:

1. The integration keep supporting both the new and old versions of truenas (pre and post electric eel)

2. The integration is migrated/updated to only support post electric eel, and we bump the major version and leave the old version to go. They seem to be moving away from the REST API, and fully migrating their [TrueNAS API Client](https://www.truenas.com/docs/scale/api/)

Personally i feel its best to go with the latter option, as its going to hinder development/improvements, but then again, it comes down to how @tomaae wants to approach it

**Comment by schlomoti0n** [comment]
> > how do i make this fix and add it back to home assistant with the fix?
> 
> You need to alter the 'coordinator.py' file in the custom components folder. If you installed using HACS, the file is located in `custom_components/truenas/coordinator.py`.
> 
> Line 1123 normally reads: ` if not self._is_scale:`
> 
> If you remove the "not" part, it should read: ` if self._is_scale:`
> 
> Then restart HA to load the modified version. It's not a fix, but it gets most stats back online until a proper fix is available (which I expect will require more changes from TrueNAS to finish the apps API on the new version).

worked fore me. Bug is still alive in TrueNAS 24.10.0

**Comment by kmanan** [comment]
> > how do i make this fix and add it back to home assistant with the fix?
> 
> You need to alter the 'coordinator.py' file in the custom components folder. If you installed using HACS, the file is located in `custom_components/truenas/coordinator.py`.
> 
> Line 1123 normally reads: ` if not self._is_scale:`
> 
> If you remove the "not" part, it should read: ` if self._is_scale:`
> 
> Then restart HA to load the modified version. It's not a fix, but it gets most stats back online until a proper fix is available (which I expect will require more changes from TrueNAS to finish the apps API on the new version).

Hi. This worked in getting most of the sensors working. now that apps have moved from a separate "tank" / k3s to docker, how can I get the app sensors working? n00b here, any ideas would be superhelpful. Thank you!

**Comment by tobse91** [comment]
> if you already update to 24.10, you can make it works by editing few lines like i did. in `custom_components/truenas/coordinator.py`, if you already edit line 1123 before, revert it back to
> 
> ```
>         if not self._is_scale:
> ```
> 
> and edit line 1128 and so on to:
> 
> ```
>             source=self.api.query("app"),
>             key="id",
>             vals=[
>                 {"name": "id", "default": 0},
>                 {"name": "name", "default": "unknown"},
>                 {"name": "human_version", "default": "unknown"},
>                 {"name": "upgrade_available", "default": "unknown"},
>                 {"name": "image_updates_available", "default": "unknown"},
>                 {"name": "portal", "source": "portals/Web UI", "default": "unknown"},
>                 {"name": "state", "default": "unknown"},
>             ],
>             ensure_vals=[
>                 {"name": "running", "type": "bool", "default": False},
>             ],
>         )
> 
>         for uid, vals in self.ds["app"].items():
>             self.ds["app"][uid]["running"] = vals["state"] == "RUNNING"
> ```
> 
> and for fixing the attibutes: in `custom_components/truenas/binary_sensor_types.py` line 78-79
> 
> ```
>     "upgrade_available",
>     "image_updates_available",
> ```
> 
> don't forget to restart homeassistant
> 
> ![image](https://private-user-images.githubusercontent.com/89054062/381809519-73364fac-a42b-4cf2-b100-8077cf696ae6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NTgyNTIsIm5iZiI6MTczMDc1Nzk1MiwicGF0aCI6Ii84OTA1NDA2Mi8zODE4MDk1MTktNzMzNjRmYWMtYTQyYi00Y2YyLWIxMDAtODA3N2NmNjk2YWU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDIyMDU1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2NmI2NjkwMjFkMjQ3NzU3YzA5ODI4YjZmNDc2MTM4NjU3MWZmYzIwMWRkNGRjMGViOWYyZWIyMWRkY2RiMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.esxILEYAk8htUsYwkEErHc5nAV-g84o9q5nqatkdwDE)

Can you merge this into a release?

**Comment by itsniawan** [comment]
> Can you suggest this for a pull request?

PR submitted with some improvement to retain compatibility with previous TrueNAS version #180

**Comment by ausghostdog** [comment]
Jsut had the same issue, that for the fix

**Comment by janusn** [comment]
Thanks @Berserkir-Wolf and @itsniawan for the temp fixes.

Hopefully a complete solution will be integrated and tagged soon.

**Comment by tomaae** [comment]
released in 1.3
